rpc/eth_api.go

```diff
@@ -430,19 +440,22 @@ func (e *PublicEthAPI) doCall(args CallArgs, blockNr rpc.BlockNumber, vmCfg vm.C
 		return []byte{}, err
 	}
 
-	var simResult sdk.Result
-	if err = ctx.Codec.UnmarshalBinaryLengthPrefixed(res, &simResult); err != nil {
-		return nil, err
-	}
-
-	_, _, ret, err := types.DecodeReturnData(simResult.Data)
-
-	return ret, err
+	return res, nil
 }
 
 // EstimateGas estimates gas usage for the given smart contract call.
-func (e *PublicEthAPI) EstimateGas(args CallArgs, blockNum BlockNumber) hexutil.Uint64 {
-	return 0
+func (e *PublicEthAPI) EstimateGas(args CallArgs, blockNr rpc.BlockNumber) (hexutil.Uint64, error) {
+	result, err := e.doCall(args, blockNr, big.NewInt(emint.DefaultRPCGasLimit))
+	if err != nil {
+		return 0, err
+	}
+
+	var simResult sdk.Result
```

Comment:
Instead of unmarshalling in the two locations, can just return a sdk.Result from the doCall instead of the encoded result to be used by both. There would be no limitations since it is just decoding the result into the proper struct.